src/lib/Type.hs

```diff
@@ -117,20 +118,17 @@ instance HasType Atom where
       zipWithM_ (|:) params paramTys
       let paramTysRemaining = drop (length params) paramTys
       return $ foldr (-->) TyKind paramTysRemaining
+    LabeledRow row -> checkLabeledRow row $> LabeledRowKind
```

Comment:
I'm a little confused as to why is this even an atom. Is it because you want to bind it as a function argument? It feels a little ugly to me, because I'm not sure if we should be reifying it as a value in user programs, since there are no constructors anyway. It's a purely syntactic construct.

Reply:
I think it needs to be an atom in order for inference to work. Suppose we see a variant expression `{| a = 4 |}`. Type inference infers that this is of type `{a : Int | ...?0}` where `?0` is an inference variable. But it's specifically an inference variable of kind `LabeledRowKind`. Now, later, when we unify this with something else, maybe we figure out that `?0 = < b: Float | ...?1 >`. We need to have some atom that we can bind `?0` to when we substitute it into our original type to yield the new type `{a:Int | b:Float | ...?1}`, so we need to have some atom whose kind is `LabeledRowKind`. (The same thing happens when we apply a function with an implicit argument of type `LabeledRowKind`.)

This happens [here](https://github.com/google-research/dex-lang/pull/201/files#diff-fe3fa5e4b62e6b351f5577e3785283f1R824), and gets substituted out [here](https://github.com/google-research/dex-lang/pull/201/files#diff-5d8f219c998ad3145f966a29811220b6R1084).

I suppose I could probably make this `neverAllowed` in `Type.hs`, though, since it's really just an artifact of inference and shouldn't exist afterward.

Reply:
It needs to be an atom so that we can reuse the atom specific inference code, but it's not a hard constraint. You could imagine generalizing it to be able to handle other types such as those labeled rows. Still, should be fine to leave for now.

Reply:
I think it also reuses the atom-specific lambda-binding substitution code even after inference (that's the implicit arg application thing I discovered when trying to make it `neverAllowed`). Probably we could make it not an atom by making other changes to how arguments work, but I'm not sure it's worth that. In any case, agreed to leave for now.